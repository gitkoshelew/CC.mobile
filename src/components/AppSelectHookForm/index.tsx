import {Control, Controller, FieldPath, FieldValues} from 'react-hook-form';
import {TextInputProps} from 'react-native';
import React from 'react';
import {AppSelect} from '@src/components/ui/AppSelect/index';
import {TextBox, BlockBox} from '@src/components/ui/ReadyStyles/Boxes/index';

type TextInputHookFormPropsType<T extends FieldValues> = TextInputProps & {
  name: FieldPath<T>;
  data: string[];
  size: 's' | 'm';
  type: 'primary' | 'secondary';
  label?: string | null;
  rules?: Object;
  control: Control<T>;
  onAppSelect?: (value: string, onPress: (value: string) => void) => void;
};

export const AppSelectHookForm = <T extends FieldValues>({
  data,
  name,
  size,
  type,
  label,
  rules = {},
  control,
  onAppSelect,
}: TextInputHookFormPropsType<T>) => {
  const handlerSelect = (value: string, onPress: (value: string) => void) => {
    onAppSelect && onAppSelect(value, onPress);
  };

  return (
    <BlockBox>
      {label && <TextBox>{label}</TextBox>}
      <Controller
        control={control}
        name={name}
        rules={rules}
        render={({field: {onChange, value}}) => (
          <AppSelect
            value={value}
            size={size}
            data={data}
            type={type}
            onSelect={(selected: string) => handlerSelect(selected, onChange)}
          />
        )}
      />
    </BlockBox>
  );
};
