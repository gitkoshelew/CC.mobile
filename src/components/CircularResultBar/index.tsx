import CircularProgress from 'react-native-circular-progress-indicator';
import {Color} from '@theme/colors';
import {StyleSheet} from 'react-native';
import {ViewCenter} from '../ui/ReadyStyles/Containers';
import {useContext} from 'react';
import {ThemeContext} from 'styled-components/native';

interface IProps {
  result: number;
}

export const CircularResultBar = ({result}: IProps) => {
  const theme = useContext(ThemeContext);

  return (
    <ViewCenter>
      <CircularProgress
        value={result}
        radius={83}
        progressValueStyle={styles.progressValueStyle}
        valueSuffix={'%'}
        duration={1500}
        activeStrokeWidth={26}
        activeStrokeColor={Color.BlueLight}
        inActiveStrokeWidth={26}
        inActiveStrokeColor={theme.box}
      />
    </ViewCenter>
  );
};

const styles = StyleSheet.create({
  progressValueStyle: {
    fontSize: 38,
    fontFamily: 'Montserrat-Regular',
    color: Color.BlueLight,
  },
});
