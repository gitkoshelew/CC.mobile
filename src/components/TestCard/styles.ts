import {StyleSheet, ViewStyle} from 'react-native';
import {DefaultTheme} from 'styled-components';

export const styles = StyleSheet.create((theme?: DefaultTheme) => ({
  container: {
    width: '100%',
    alignItems: 'center',
  } as ViewStyle,
  test: {
    width: '90%',
    flexDirection: 'row',
    paddingHorizontal: 15,
    backgroundColor: theme?.block,
    height: 71,
    marginVertical: 10,
    justifyContent: 'space-between',
    alignItems: 'center',
    borderRadius: 10,
    shadowOpacity: 0.1,
    shadowOffset: {
      width: 0,
      height: 20,
    },
    shadowRadius: 10,
    elevation: 5,
  } as ViewStyle,
  title: {
    fontSize: 16,
    color: theme?.textMainColor,
    fontWeight: '600',
  } as ViewStyle,
  description: {
    fontSize: 14,
    color: theme?.textMainColor,
    fontWeight: '500',
  } as ViewStyle,
  contentContainer: {
    marginTop: 6,
    justifyContent: 'space-between',
  } as ViewStyle,
  iconTopic: {
    marginRight: 5,
    marginTop: 3,
    marginLeft: 2,
  },
  icon: {
    marginRight: 5,
    marginTop: 2,
  },
  text: {
    color: theme?.textMainColor,
    fontFamily: 'Montserrat-Regular',
    fontSize: 13,
  },
  content: {
    flexDirection: 'row',
  } as ViewStyle,
}));
