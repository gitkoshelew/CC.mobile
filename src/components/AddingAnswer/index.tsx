import {TouchableOpacity, View} from 'react-native';
import {styles} from './styles';
import {BlockAnswerBox} from '../ui/ReadyStyles/Boxes';
import {CustomTextInput} from '../ui/CustomTextInput';
import {CheckBox} from '../ui/CheckBox';
import AntDesign from 'react-native-vector-icons/AntDesign';
import {Color} from '@theme/colors';
import {useCallback, useEffect, useState} from 'react';
import {Control, Controller, useWatch} from 'react-hook-form';
import {useTranslation} from 'react-i18next';
import {CreateQuestionFieldType} from '@src/screens/CreateQuiz/components/CreateQuestion/CreateQuestion';

type AddingAnswerPropsType = {
  type: string;
  index: number;
  option: string;
  control: Control<CreateQuestionFieldType>;
  correctAnswer: string[];
  onPressDelete: (index: number) => void;
  isCheckingDuplicate: boolean;
  isDisabledDeleteBtn: boolean;
  onPressCorrectAnswer: (index: number, checked: boolean, textOption: string) => void;
};

export const AddingAnswer = (props: AddingAnswerPropsType) => {
  const {
    type,
    index,
    option,
    control,
    onPressDelete,
    correctAnswer,
    isCheckingDuplicate,
    isDisabledDeleteBtn,
    onPressCorrectAnswer,
  } = props;
  const {t} = useTranslation(['validationFields']);
  const findCorrectInput = correctAnswer.find(el => el === option);
  const [isChecked, setIsChecked] = useState(false);
  const [correctInput, setCorrectInput] = useState(findCorrectInput);
  const isCurrentOptionText = useWatch({
    name: `options.${index}.option`,
    control,
  });

  const onPressCorrectAnswerHandler = useCallback(
    (checked: boolean) => {
      setIsChecked(checked);
      setCorrectInput(isCurrentOptionText);
      onPressCorrectAnswer(index, checked, isCurrentOptionText);
    },
    [index, isCurrentOptionText, onPressCorrectAnswer],
  );
  const onPressDeletePressed = useCallback(() => {
    onPressDelete(index);
  }, [index, onPressDelete]);

  useEffect(() => {
    if (correctInput !== isCurrentOptionText) {
      onPressCorrectAnswer(index, false, correctInput!);
      setIsChecked(false);
      setCorrectInput('');
    }
  }, [index, correctInput, isCurrentOptionText, onPressCorrectAnswer]);

  useEffect(() => {
    setIsChecked(correctAnswer.includes(isCurrentOptionText));
  }, [isCurrentOptionText, correctAnswer, option]);

  const isSingle = type === 'single';
  const isSingleDisabled = (!isChecked && correctAnswer.length) || !isCurrentOptionText;
  const isMultipleDisabled = !isCurrentOptionText;

  const isDisabledDeleteOption = isDisabledDeleteBtn || isChecked;

  return (
    <BlockAnswerBox>
      <View style={styles.inputBox}>
        <Controller
          render={({field: {onChange, value, onBlur}, fieldState: {error}}) => (
            <CustomTextInput
              onChangeText={onChange}
              value={value}
              onBlur={onBlur}
              error={error}
            />
          )}
          rules={{
            required: `${t('option.required')}`,
            maxLength: {
              value: 50,
              message: `${t('validationFields.option.minLength')}`,
            },
          }}
          name={`options.${index}.option`}
          control={props.control}
        />
      </View>
      <View style={styles.picker}>
        <CheckBox
          onPress={onPressCorrectAnswerHandler}
          isChecked={isChecked}
          disabled={Boolean(isSingle ? isSingleDisabled : isMultipleDisabled)}
          fillColor={isCheckingDuplicate ? Color.Red : Color.GreenLight}
        />
        <TouchableOpacity
          style={isDisabledDeleteOption && styles.disabled}
          onPress={onPressDeletePressed}
          disabled={isDisabledDeleteOption}>
          <AntDesign name="minuscircleo" size={30} color={Color.Red} />
        </TouchableOpacity>
      </View>
    </BlockAnswerBox>
  );
};
