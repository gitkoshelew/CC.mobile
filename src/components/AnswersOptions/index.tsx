import {useContext} from 'react';
import {View} from 'react-native';
import RadioForm, {RadioButtonInput, RadioButtonLabel} from 'react-native-simple-radio-button';
import {AnswerRadioContainer, styles, ViewMarginRight} from './styles';
import {Color} from '@theme/colors';
import {MultipleCheckboxes} from '@src/components/MultipleCheckboxes/MultipleCheckboxes';
import {CustomText} from '@src/components/ui/ReadyStyles/Boxes/index';
import {ThemeContext} from 'styled-components/native';

type AnswersOptionsPropsType = {
  onPress: (label: string, value: number) => void;
  data: string[];
  answerType: string;
  selected?: number;
  dataOptions: IDataOptions[];
  onPressCheck: (label: string, value: number, isChecked: boolean) => void;
};
export type IDataOptions = {
  label: string;
  value: number;
  isChecked: boolean;
};

export const AnswersOptions = ({
  onPress,
  onPressCheck,
  dataOptions,
  data,
  selected,
  answerType,
}: AnswersOptionsPropsType) => {
  const theme = useContext(ThemeContext);

  const onPressRadio = (value: number) => {
    let answer: string = data[value];
    onPress(answer, value);
  };
  const handlerPressCheck = (label: string, value: number, isChecked: boolean) => {
    onPressCheck(label, value, isChecked);
  };

  return (
    <View>
      <RadioForm formHorizontal={false} animation={true}>
        {dataOptions.map((obj, i) => (
          <AnswerRadioContainer key={i}>
            <ViewMarginRight>
              {answerType === 'single' ? (
                <RadioButtonInput
                  obj={obj}
                  index={i}
                  isSelected={selected === i}
                  onPress={onPressRadio}
                  buttonOuterColor={selected === i ? Color.BlueLight : Color.GrayStrongDark}
                  buttonSize={15}
                  buttonOuterSize={30}
                />
              ) : (
                <MultipleCheckboxes item={obj} onPress={handlerPressCheck} />
              )}
            </ViewMarginRight>
            {answerType === 'single' ? (
              <RadioButtonLabel
                obj={obj}
                index={i}
                labelHorizontal={false}
                onPress={onPressRadio}
                labelStyle={styles(theme).textRadio}
              />
            ) : (
              <CustomText key={i} fs="16px">
                {obj.label}
              </CustomText>
            )}
          </AnswerRadioContainer>
        ))}
      </RadioForm>
    </View>
  );
};
