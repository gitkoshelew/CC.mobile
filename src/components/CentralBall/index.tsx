import React from 'react';
import {Color} from '@theme/colors';
import {Text} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import {styles} from '@src/components/CentralBall/styles';
import {ViewDynamicFlex} from '@src/components/ui/ReadyStyles/Containers/index';

type CentralBallPropsType = {
  value: number;
  description: string;
};

export const CentralBall = ({value, description}: CentralBallPropsType) => {
  return (
    <LinearGradient
      start={{x: 0.0, y: 0.05}}
      end={{x: 0.6, y: 1.0}}
      colors={[Color.BlueLight, Color.DarkBlue]}
      style={styles.container}>
      <ViewDynamicFlex alignI="center" justifyC="center">
        <Text style={styles.value}>{value}</Text>
        <Text style={styles.description}>{description}</Text>
      </ViewDynamicFlex>
    </LinearGradient>
  );
};
