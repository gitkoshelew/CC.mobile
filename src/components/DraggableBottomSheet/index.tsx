import {useEffect, useRef, useState} from 'react';
import {
  Animated,
  Dimensions,
  ImageBackground,
  PanResponder,
  ScaledSize,
  View,
} from 'react-native';
import {Content} from '@src/components/DraggableBottomSheet/Content/Content';
import {useAppDispatch, useAppSelector} from '@hooks/hooks';
import {getStyles} from '@src/components/DraggableBottomSheet/styles';
import {setIsScrollEnabled} from '@src/bll/appReducer';

export const {width: WINDOW_WIDTH, height: WINDOW_HEIGHT}: ScaledSize =
  Dimensions.get('window');

const BOTTOM_SHEET_MIN_HEIGHT = WINDOW_HEIGHT * 0.03;
const MAX_DOWNWARD_TRANSLATE_Y = 0;
const DRAG_THRESHOLD = 30;

export const DraggableBottomSheet = () => {
  const dispatch = useAppDispatch();
  const [isOpen, setIsOpen] = useState(false);
  const [isLogForm, setIsLogForm] = useState(false);

  const isLoggedIn = useAppSelector(state => state.authReducer.isLoggedIn);

  const BOTTOM_SHEET_MAX_HEIGHT = isLoggedIn
    ? WINDOW_HEIGHT * 0.26
    : isLogForm
    ? WINDOW_HEIGHT * 0.5
    : WINDOW_HEIGHT * 0.68;
  const MAX_UPWARD_TRANSLATE_Y = BOTTOM_SHEET_MIN_HEIGHT - BOTTOM_SHEET_MAX_HEIGHT;

  const animatedValue = useRef(new Animated.Value(0)).current;
  const lastGestureDy = useRef(0);
  const panResponder = useRef(
    PanResponder.create({
      onStartShouldSetPanResponder: () => true,
      onPanResponderGrant: () => {
        dispatch(setIsScrollEnabled(false));
        animatedValue.setOffset(lastGestureDy.current);
      },
      onPanResponderEnd: () => {
        dispatch(setIsScrollEnabled(true));
      },
      onPanResponderMove: (e, gesture) => {
        animatedValue.setValue(gesture.dy);
      },
      onPanResponderRelease: (e, gesture) => {
        animatedValue.flattenOffset();
        if (gesture.dy > 0) {
          if (gesture.dy <= DRAG_THRESHOLD) {
            springAnimation('up');
          } else {
            springAnimation('down');
          }
        } else {
          if (gesture.dy >= -DRAG_THRESHOLD) {
            springAnimation('down');
          } else {
            springAnimation('up');
          }
        }
      },
    }),
  ).current;

  const springAnimation = (direction: 'up' | 'down') => {
    lastGestureDy.current =
      direction === 'down' ? MAX_DOWNWARD_TRANSLATE_Y : MAX_UPWARD_TRANSLATE_Y;
    Animated.spring(animatedValue, {
      toValue: lastGestureDy.current,
      useNativeDriver: true,
    }).start();
    lastGestureDy.current >= 0 ? setIsOpen(false) : setIsOpen(true);
  };

  const bottomSheetAnimation = {
    transform: [
      {
        translateY: animatedValue.interpolate({
          inputRange: [MAX_UPWARD_TRANSLATE_Y, MAX_DOWNWARD_TRANSLATE_Y],
          outputRange: [MAX_UPWARD_TRANSLATE_Y, MAX_DOWNWARD_TRANSLATE_Y],
          extrapolate: 'clamp',
        }),
      },
    ],
  };

  useEffect(() => {
    !isLoggedIn && setIsLogForm(true);
  }, [isLoggedIn]);

  return (
    <View style={getStyles().container}>
      <Animated.View
        style={[
          getStyles(BOTTOM_SHEET_MAX_HEIGHT, MAX_UPWARD_TRANSLATE_Y).bottomSheet,
          bottomSheetAnimation,
        ]}>
        <ImageBackground
          style={getStyles().imageBackground}
          source={require('../../assets/images/background-second.png')}
          resizeMode="stretch">
          <View style={getStyles().draggableArea} {...panResponder.panHandlers}>
            <View style={getStyles().dragHandle} />
          </View>
          <Content isOpen={isOpen} isLogForm={isLogForm} setIsLogForm={setIsLogForm} />
        </ImageBackground>
      </Animated.View>
    </View>
  );
};
