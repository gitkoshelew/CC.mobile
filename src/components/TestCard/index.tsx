import {Text, View} from 'react-native';
import {styles} from './styles';
import {AppButton} from '../ui/AppButton';
import {TypeAppButton} from '@customTypes/AppButtun-types';
import Fontisto from 'react-native-vector-icons/Fontisto';
import {Color} from '@theme/colors';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import {useContext} from 'react';
import {ThemeContext} from 'styled-components/native';

type ITestCard = {
  id: number;
  questions: number;
  topic: string;
  onPress: (id: number, questions: number) => void;
  title: string;
};

export const TestCard = ({onPress, title, id, topic, questions}: ITestCard) => {
  const theme = useContext(ThemeContext);

  return (
    <View style={styles().container}>
      <View style={styles(theme).test}>
        <View>
          <Text style={styles(theme).title}>{title}</Text>
          <View style={styles().contentContainer}>
            <View style={styles().content}>
              <Fontisto
                style={styles().iconTopic}
                name="hashtag"
                color={Color.BlueLight}
                size={10}
              />
              <Text style={styles(theme).text}>{topic}</Text>
            </View>
            <View style={styles().content}>
              <MaterialCommunityIcons
                style={styles().icon}
                name="progress-question"
                color={Color.BlueLight}
                size={13}
              />
              <Text style={styles(theme).text}>difficulty</Text>
            </View>
          </View>
        </View>
        <Text style={styles(theme).description}>Questions: {questions}</Text>
        <AppButton
          type={TypeAppButton.PRIMARY}
          title="Start"
          onPress={() => onPress(id, questions)}
        />
      </View>
    </View>
  );
};
