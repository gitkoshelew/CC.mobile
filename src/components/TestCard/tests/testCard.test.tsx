import {fireEvent, render} from '@testing-library/react-native';
import {TestCard} from '@src/components/TestCard';

test('Should render testCard list', async () => {
  const {getByText, toJSON} = render(
    <TestCard onPress={() => {}} id={1} title={'Star'} questions={1} topic="JS" />,
  );
  const button = getByText('Start');
  fireEvent.press(button);
  expect(toJSON()).toMatchSnapshot();
});
