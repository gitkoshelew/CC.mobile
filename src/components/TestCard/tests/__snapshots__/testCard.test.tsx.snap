// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render testCard list 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": undefined,
        "borderRadius": 10,
        "elevation": 5,
        "flexDirection": "row",
        "height": 71,
        "justifyContent": "space-between",
        "marginVertical": 10,
        "paddingHorizontal": 15,
        "shadowOffset": {
          "height": 20,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 10,
        "width": "90%",
      }
    }
  >
    <View>
      <Text
        style={
          {
            "color": undefined,
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
      >
        Star
      </Text>
      <View
        style={
          {
            "justifyContent": "space-between",
            "marginTop": 6,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#6B9DE9",
                  "fontSize": 10,
                },
                {
                  "marginLeft": 2,
                  "marginRight": 5,
                  "marginTop": 3,
                },
                {
                  "fontFamily": "Fontisto",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            
          </Text>
          <Text
            style={
              {
                "color": undefined,
                "fontFamily": "Montserrat-Regular",
                "fontSize": 13,
              }
            }
          >
            JS
          </Text>
        </View>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#6B9DE9",
                  "fontSize": 13,
                },
                {
                  "marginRight": 5,
                  "marginTop": 2,
                },
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󱔢
          </Text>
          <Text
            style={
              {
                "color": undefined,
                "fontFamily": "Montserrat-Regular",
                "fontSize": 13,
              }
            }
          >
            difficulty
          </Text>
        </View>
      </View>
    </View>
    <Text
      style={
        {
          "color": undefined,
          "fontSize": 14,
          "fontWeight": "500",
        }
      }
    >
      Questions: 
      1
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#2D3E6B",
            "borderRadius": 20,
            "height": 38,
            "justifyContent": "center",
            "opacity": 1,
            "paddingHorizontal": 14,
          }
        }
      >
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontFamily": "Montserrat-Regular",
              "fontSize": 14,
            }
          }
        >
          Start
        </Text>
      </View>
    </View>
  </View>
</View>
`;
